subsys/lorawan/nvm/lorawan_nvm_settings.c

```diff
@@ -0,0 +1,202 @@
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_MAC_GROUP1, MacGroup1)
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_MAC_GROUP2, MacGroup2)
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_SECURE_ELEMENT,
+		      SecureElement)
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_REGION_GROUP1, RegionGroup1)
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_REGION_GROUP2, RegionGroup2)
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_CLASS_B, ClassB)
+
+	settings_save();
+	/* Clean notification flag */
+	nvm_notify_flag = LORAMAC_NVM_NOTIFY_FLAG_NONE;
+}
+
+void lorawan_nvm_data_mgmt_event(uint16_t notifyFlags)
+{
+	LOG_WRN("New NVM notify flag: 0x%04X", notifyFlags);
+
+	nvm_notify_flag = notifyFlags;
+
+	if (notifyFlags != LORAMAC_NVM_NOTIFY_FLAG_NONE) {
+		lorawan_nvm_save_settings();
+	}
+}
+
+static int load_setting(void *tgt, size_t tgt_size,
+			const char *key, size_t len,
+			settings_read_cb read_cb, void *cb_arg)
+{
+	if (len != tgt_size) {
+		LOG_ERR("Can't load '%s' state, size mismatch.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	if (!tgt) {
+		LOG_ERR("Can't load '%s' state, no target.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	if (read_cb(cb_arg, tgt, len) != len) {
+		LOG_ERR("Can't load '%s' state, short read.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+#define LOAD_SETTINGS(nvm, name)						  \
+({										  \
+int $err = -EAGAIN;								  \
+if (strcmp(#name, key) == 0) {						  \
+	$err = load_setting(&(nvm)->name,					  \
+			   sizeof((nvm)->name),				  \
+			   key, len, read_cb, cb_arg);			  \
+	if ($err) {								  \
+		LOG_ERR("Could not read settings" # name);			  \
+	}									  \
+	$err = 0;								  \
+}										  \
+$err;										  \
+})
+
+static int on_setting_loaded(const char *key, size_t len,
+			   settings_read_cb read_cb,
+			   void *cb_arg, void *param)
+{
+	int err;
+	LoRaMacNvmData_t *nvm = param;
+
+	LOG_DBG("Key: %s", log_strdup(key));
+
+	err = LOAD_SETTINGS(nvm, Crypto);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, MacGroup1);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, MacGroup2);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, SecureElement);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, RegionGroup1);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, RegionGroup2);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	err = LOAD_SETTINGS(nvm, ClassB);
+	if (err != -EAGAIN) {
+		return err;
+	}
+
+	LOG_WRN("Unknown LoRaWAN setting: %s", log_strdup(key));
+	return 0;
+}
+
+int lorawan_nvm_data_restore(void)
+{
+	int err;
+	LoRaMacStatus_t status;
+	MibRequestConfirm_t mibReq;
+
+	/* Retrieve the actual context */
+	mibReq.Type = MIB_NVM_CTXS;
+	if (LoRaMacMibGetRequestConfirm(&mibReq) != LORAMAC_STATUS_OK) {
+		LOG_ERR("Could not get NVM context");
+		return -EINVAL;
+	}
+
+	err = settings_load_subtree_direct(LORAWAN_SETTINGS_BASE,
```

Comment:
It's possible (likely even) that this runs before the settings subsystem has been initialized. Even if an application were to call `settings_subsys_init()` in `main()` it would be too late since this code runs during initialization.

Reply:
Now, I remember why I removed lorawan_init() from SYS_INIT. That's the problem I've faced when testing the NVM implementation.
Should we change lorawan_init() visibility to public, and require to call it in the main, after settings_load() has been called?

Reply:
settings_subsys_init() is safe to call multiple times, and can be called in lorawan_init() inside the conditional right before lorawan_nvm_data_restore() is called